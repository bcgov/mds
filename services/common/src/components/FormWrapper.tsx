import { Button, Row, Typography } from "antd";
import React from "react";

export const { Provider: FormProvider, Consumer: FormConsumer } = React.createContext({
  view: false,
});

interface FormWrapperProps {
  name: string;
  onSubmit: any;
  children: any;
  showFooter?: boolean;
  okText?: string;
  onOk?: any;
  cancelText?: string;
  onCancel?: any;
  loading?: boolean;
  view?: boolean;
}
// why not antd v4 form? The submit doesn't work with redux.
const FormWrapper = ({
  name,
  onSubmit,
  children,
  showFooter = true,
  okText = "Submit",
  cancelText = "Cancel",
  onOk = () => {},
  onCancel = () => {},
  loading = false,
  view = false,
}: FormWrapperProps) => {
  const getFooter = () => {
    if (!showFooter || view) {
      return null;
    }
    return (
      <Row className="flex-between form-button-container-row">
        {onCancel && (
          <Button className="full-mobile" type="ghost" onClick={onCancel}>
            {cancelText}
          </Button>
        )}
        {(onOk || okText) && (
          <Button
            onClick={onOk || null}
            loading={loading}
            type="primary"
            className="full-mobile"
            htmlType="submit"
          >
            {okText}
          </Button>
        )}
      </Row>
    );
  };

  return (
    <FormProvider value={{ view }}>
      <form id={name} name={name} className="ant-form ant-form-vertical" onSubmit={onSubmit}>
        {children}
        {getFooter()}
      </form>
    </FormProvider>
  );
};

// general use for showing data
export const ViewFormDataItem = (label = "", value: string | number = "") => {
  return (
    <>
      {label.length && <Typography.Paragraph strong>{label}</Typography.Paragraph>}
      <Typography.Paragraph>{value}</Typography.Paragraph>
    </>
  );
};

// inputs should extend (if necessary) same base props type
export interface BaseInputProps {
  label: string;
  id: string;
  defaultValue?: any;

  placeholder?: string;
  meta?: any;
  input?: any;
  disabled?: boolean;
  allowClear?: boolean;
  required?: boolean;
}

export default FormWrapper;
