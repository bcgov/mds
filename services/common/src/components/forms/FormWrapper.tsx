import React, { FC, useEffect } from "react";
import { Form } from "antd";
import { compose } from "@reduxjs/toolkit";
import { connect, useDispatch, useSelector } from "react-redux";
import { reduxForm, submit, getFormSubmitErrors, InjectedFormProps, ConfigProps } from "redux-form";

export interface IFormContext {
  isEditMode: boolean;
  isModal: boolean;
}
/**
 * The values in FormProvider (from FormWrapper props) will be passed down to child components
 * without having to pass them down through the whole tree, manually
 * Import FormConsumer to access these values within an input component
 * More attributes can be added if necessary, but it should be kept minimal
 */
export const { Provider: FormProvider, Consumer: FormConsumer } = React.createContext<IFormContext>(
  {
    isEditMode: true,
    isModal: false,
  }
);

// FormWrapper EXAMPLE USAGE:
/** 
export const MyForm = () => {
  return (
    <FormWrapper 
      name="my_form"
      onSubmit={(values) => doSomething(values)}
      isEditMode={booleanVariable}
      reduxFormConfig={{touchOnChange: true}}
    >
      <Field 
        name="field_name"
        label="Field Name"
        props={{ mySpecificAttribute: someData }} 
        required 
        validate={[required]} 
        component={RenderMyInput}
      />
      ... more Fields
      <Button htmlType="submit">Submit</Button>
    </FormWrapper>
  );
}
PROPS:
- anything in reduxFormConfig will get passed to the reduxForm constructor
- the name of the form is necessary to connect to the store properly
- onSubmit will only be called when there are no submit errors (validation passes)

NOTABLE OMISSIONS:
- 'MyForm' should not include reduxForm, ant design Form or Form.Item

SEE ALSO:
- BaseInput.tsx 
*/
interface FormWrapperProps {
  name: string;
  initialValues?: any;
  reduxFormConfig?: Partial<ConfigProps>;
  onSubmit: (values) => void | Promise<void>;
  children: any;
  isModal?: boolean;
  loading?: boolean;
  isEditMode?: boolean;
  scrollOnToggleEdit?: boolean;
}

const FormWrapper: FC<FormWrapperProps & InjectedFormProps<any>> = ({
  isEditMode = true,
  isModal = false,
  scrollOnToggleEdit = true,
  children,
  ...props
}) => {
  const providerValues = {
    isEditMode,
    isModal,
  };
  const dispatch = useDispatch();
  const formErrors = useSelector(getFormSubmitErrors(props.name));

  useEffect(() => {
    if (scrollOnToggleEdit) {
      window.scrollTo(0, 0);
    }
  }, [isEditMode]);

  const handleSubmit = (values) => {
    dispatch(submit(props.name));
    if (!formErrors) {
      props.onSubmit(values);
    }
  };

  const formClassName =
    `common-form common-form-${props.name} form-` + isEditMode ? "edit" : "view";

  return (
    <FormProvider value={providerValues}>
      <Form layout="vertical" onFinish={handleSubmit} name={props.name} className={formClassName}>
        {children}
      </Form>
    </FormProvider>
  );
};

const mapStateToProps = (_state, ownProps) => ({
  form: ownProps.name,
  initialValues: ownProps.initialValues,
  ...ownProps.reduxFormConfig,
});
export default compose(connect(mapStateToProps), reduxForm({}))(FormWrapper as any) as FC<
  FormWrapperProps
>;
