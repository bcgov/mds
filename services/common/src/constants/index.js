/** This is the source of the eslint import errors.
 * These files contain many exports with identical names.
 * This will need some cleaning, but the commented out files are
 * not currently being imported anywhere, so this will at least
 * stop the errors for now.
 */
// export * from "./actionTypes";
// export * from "./API";
// export * from "./reducerTypes";
export * from "./environment";
export * from "./strings";
export * from "./fileTypes";
export * from "./utils";
