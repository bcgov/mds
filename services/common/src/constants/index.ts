/** This is the source of the eslint import errors.
 * These files contain many exports with identical names.
 * We'll need to fix this since the system will have no idea which
 * export to use.  It'll likely require some renaming since it seems
 * yarn workspaces requires the export to come in the form of an index
 * as is currently set up.
 */
export * from "./actionTypes";
export * from "./API";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export * from "./reducerTypes";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export * from "./strings";
export * from "./environment";
export * from "./fileTypes";
export * from "./utils";
export * from "./enums";
export * from "./forms";
