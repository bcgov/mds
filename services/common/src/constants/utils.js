export const serverSidePaginationOptions = (pageData) => {
  return {
    defaultCurrent: 1,
    current: pageData?.current,
    total: pageData?.total,
    pageSize: pageData?.items_per_page,
    showSizeChanger: true,
    position: ["bottomCenter"],
  };
};

export const parseSortDir = (sortDir) => {
  if (!sortDir) return undefined;
  return sortDir === "ascend" ? "asc" : "desc";
};

export const parseServerSideSearchOptions = (pagination, filters, sorter) => {
  Object.keys(filters).forEach((key) => !filters[key] && delete filters[key]);
  return {
    per_page: pagination.pageSize,
    page: pagination?.current,
    sort_field: sorter?.column?.dataIndex || undefined,
    sort_dir: parseSortDir(sorter.order),
    ...filters,
  };
};

export const removeNullValues = (obj) => {
  Object.keys(obj).forEach((key) => obj[key] === null && delete obj[key]);
  return obj;
};

export const getStatusDescription = (
  projectSummaryStatusCode,
  majorMineApplicationStatusCode,
  irtStatusCode
) => {
  if (
    projectSummaryStatusCode === "WDN" ||
    majorMineApplicationStatusCode === "WDN" ||
    irtStatusCode === "WDN"
  ) {
    return "Inactive";
  }
  return "Active";
};
