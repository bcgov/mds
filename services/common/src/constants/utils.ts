import { SystemFlagEnum } from "./enums";

export const serverSidePaginationOptions = (pageData) => {
  return {
    defaultCurrent: 1,
    current: pageData?.current,
    total: pageData?.total,
    pageSize: pageData?.items_per_page,
    showSizeChanger: true,
    position: ["bottomCenter"],
  };
};

export const parseSortDir = (sortDir) => {
  if (!sortDir) return undefined;
  return sortDir === "ascend" ? "asc" : "desc";
};

export const parseServerSideSearchOptions = (pagination, filters, sorter) => {
  Object.keys(filters).forEach((key) => !filters[key] && delete filters[key]);
  return {
    per_page: pagination.pageSize,
    page: pagination?.current,
    sort_field: sorter?.column?.dataIndex || undefined,
    sort_dir: parseSortDir(sorter.order),
    ...filters,
  };
};

export const removeNullValues = (obj) => {
  Object.keys(obj).forEach((key) => obj[key] === null && delete obj[key]);
  return obj;
};

export const removeNullValuesRecursive = (values: any) => {
  // spread to prevent unwanted state mutation
  const obj = { ...values };
  Object.keys(obj).forEach((key) => {
    const val = obj[key];
    if (val === null) {
      delete obj[key];
      // if the property is an object, allow empty array
    } else if (val instanceof Object === true && !Array.isArray(val)) {
      const temp = removeNullValuesRecursive(val);
      // and if none left, delete it
      if (Object.keys(temp).length === 0) {
        delete obj[key];
      }
    }
  });
  return obj;
};

export const getProjectStatusDescription = (
  projectSummaryStatusCode,
  majorMineApplicationStatusCode,
  irtStatusCode
) => {
  if (
    projectSummaryStatusCode === "WDN" ||
    majorMineApplicationStatusCode === "WDN" ||
    irtStatusCode === "WDN"
  ) {
    return "Inactive";
  }
  return "Active";
};

export const isFieldDisabled = (
  systemFlag,
  projectSummaryStatusCode = "",
  isAuthorizationPage = false
) => {
  // Return false if status = "" => "Not Started"
  if (!projectSummaryStatusCode) return false;

  const disabledStatuses = new Set(["WDN", "COM"]);
  const enabledStatusMapping = {
    [SystemFlagEnum.ms]: new Set(["DFT", "CHR"]),
    [SystemFlagEnum.core]: new Set(["ASG", "UNR", "CHR", "OHD"]),
  };

  if (disabledStatuses.has(projectSummaryStatusCode)) return true;

  const enabledStatuses = enabledStatusMapping[systemFlag];
  const isFieldEnabled = enabledStatuses?.has(projectSummaryStatusCode) ?? false;

  if (!isFieldEnabled) return true;

  return isAuthorizationPage ? projectSummaryStatusCode !== "DFT" : false;
};

export const isDocumentFieldDisabled = (systemFlag, projectSummaryStatusCode = "") => {
  // Return false if status = "" => "Not Started"
  if (!projectSummaryStatusCode) return false;

  const disabledStatuses = new Set(["WDN", "COM"]);
  const enabledStatusMapping = {
    [SystemFlagEnum.ms]: new Set(["DFT", "SUB", "ASG", "CHR"]),
    [SystemFlagEnum.core]: new Set(["ASG", "UNR", "CHR", "OHD"]),
  };

  if (disabledStatuses.has(projectSummaryStatusCode)) return true;

  const enabledStatuses = enabledStatusMapping[systemFlag];
  return !enabledStatuses?.has(projectSummaryStatusCode);
};
