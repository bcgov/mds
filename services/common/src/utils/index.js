export * from "./actionlessNetworkCalls";
export * from "./environmentUtils";
export * from "./helpers";
export * from "./RequestHeaders";
export * from "./trackers";
export * from "./Validate";
