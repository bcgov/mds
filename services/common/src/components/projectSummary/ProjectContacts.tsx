import React, { FC, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { isNil } from "lodash";
import { Typography, Button, Row, Col, Popconfirm } from "antd";
import DeleteOutlined from "@ant-design/icons/DeleteOutlined";
import PlusOutlined from "@ant-design/icons/PlusOutlined";
import { Field, FieldArray, arrayPush, getFormValues, change } from "redux-form";
import {
  maxLength,
  phoneNumber,
  required,
  email,
  postalCodeWithCountry,
} from "@mds/common/redux/utils/Validate";
import { normalizePhone } from "@mds/common/redux/utils/helpers";
import LinkButton from "@mds/common/components/common/LinkButton";
import { FORM } from "@mds/common/constants/forms";
import RenderField from "@mds/common/components/forms/RenderField";
import RenderSelect from "@mds/common/components/forms/RenderSelect";
import { CONTACTS_COUNTRY_OPTIONS } from "@mds/common/constants";
import { getDropdownProvinceOptions } from "@mds/common/redux/selectors/staticContentSelectors";

const RenderContacts = (props) => {
  const { fields, contacts } = props;
  const dispatch = useDispatch();
  const provinceOptions = useSelector(getDropdownProvinceOptions);
  const handleClearProvince = (currentCountry, addressTypeCode, subDivisionCode, field) => {
    // clear out the province if country has changed and it no longer matches
    if (addressTypeCode !== currentCountry) {
      const selectedProvince = currentCountry
        ? provinceOptions.find((p) => p.value === subDivisionCode)
        : {};

      const contactFieldToBeChanged = `${field}.address.sub_division_code`;
      if (addressTypeCode === "INT" || selectedProvince?.subType !== currentCountry) {
        dispatch(change(FORM.ADD_EDIT_PROJECT_SUMMARY, contactFieldToBeChanged, null));
      }
    }
  };

  return (
    <>
      {fields.map((field, index) => {
        const { address = {} } = contacts[index] ?? {};
        const { address_type_code, sub_division_code } = address ?? {};
        const isInternational = address_type_code === "INT";
        const isPrimary = contacts[index].is_primary;
        return (
          // eslint-disable-next-line react/no-array-index-key
          <div key={index}>
            {index === 0 ? (
              <>
                <Typography.Title level={5}>Primary project contact</Typography.Title>
                <Typography.Paragraph>
                  Provide contact information for the person who has the main responsibility for
                  coordinating this project.
                </Typography.Paragraph>
              </>
            ) : (
              <>
                <Row gutter={16}>
                  <Col span={10}>
                    <Typography.Title level={5}>
                      Additional project contact #{index}
                    </Typography.Title>
                  </Col>
                  <Col span={12}>
                    <Popconfirm
                      placement="topLeft"
                      title="Are you sure you want to remove this contact?"
                      onConfirm={() => fields.remove(index)}
                      okText="Remove"
                      cancelText="Cancel"
                    >
                      <Button type="primary" size="small" ghost>
                        <DeleteOutlined className="padding-sm--left icon-sm" />
                      </Button>
                    </Popconfirm>
                  </Col>
                </Row>
              </>
            )}
            <Row gutter={16}>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.first_name`}
                  label="First Name"
                  component={RenderField}
                  required
                  validate={[required, maxLength(100)]}
                />
              </Col>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.last_name`}
                  label="Last Name"
                  component={RenderField}
                  required
                  validate={[required, maxLength(100)]}
                />
              </Col>
            </Row>

            <Row gutter={16}>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.job_title`}
                  id={`${field}.job_title`}
                  label="Job Title"
                  component={RenderField}
                />
              </Col>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.company_name`}
                  id={`${field}.company_name`}
                  label="Company Name"
                  component={RenderField}
                />
              </Col>
            </Row>

            <Row gutter={16}>
              <Col md={8} sm={19}>
                <Field
                  name={`${field}.phone_number`}
                  id={`${field}.phone_number`}
                  label="Contact Number"
                  required
                  component={RenderField}
                  validate={[phoneNumber, maxLength(12), required]}
                  normalize={normalizePhone}
                />
              </Col>
              <Col md={4} sm={5}>
                <Field
                  name={`${field}.phone_extension`}
                  id={`${field}.phone_extension`}
                  label="Ext."
                  component={RenderField}
                  validate={[maxLength(6)]}
                />
              </Col>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.email`}
                  id={`${field}.email`}
                  label="Email"
                  required
                  component={RenderField}
                  validate={[required, email]}
                />
              </Col>
            </Row>

            <Typography.Title level={5}>Mailing Address</Typography.Title>
            <Row gutter={16}>
              <Col md={19} sm={24}>
                <Field
                  name={`${field}.address.address_line_1`}
                  label="Street"
                  required={isPrimary}
                  validate={isPrimary ? [required] : []}
                  component={RenderField}
                />
              </Col>
              <Col md={5} sm={24}>
                <Field name={`${field}.address.suite_no`} label="Unit #" component={RenderField} />
              </Col>
            </Row>

            <Row gutter={16}>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.address.address_type_code`}
                  label="Country"
                  required={isPrimary}
                  validate={isPrimary ? [required] : []}
                  data={CONTACTS_COUNTRY_OPTIONS}
                  component={RenderSelect}
                  onSelect={(e) =>
                    handleClearProvince(e, address_type_code, sub_division_code, field)
                  }
                />
              </Col>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.address.sub_division_code`}
                  label="Province"
                  required={isPrimary && !isInternational}
                  data={provinceOptions.filter((p) => p.subType === address_type_code)}
                  validate={!isInternational && isPrimary ? [required] : []}
                  component={RenderSelect}
                />
              </Col>
            </Row>

            <Row gutter={16}>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.address.city`}
                  label="City"
                  required={isPrimary}
                  validate={isPrimary ? [required] : []}
                  component={RenderField}
                />
              </Col>
              <Col md={12} sm={24}>
                <Field
                  name={`${field}.address.post_code`}
                  label="Postal Code"
                  component={RenderField}
                  required={isPrimary}
                  validate={[postalCodeWithCountry(address_type_code), maxLength(10)]}
                />
              </Col>
            </Row>

            {index === 0 && (
              <>
                <Typography.Title level={5}>
                  Additional project contacts (optional)
                </Typography.Title>
                <Typography.Paragraph>
                  Provide contact information for additional people we can contact about this
                  project.
                </Typography.Paragraph>
              </>
            )}
          </div>
        );
      })}
      <LinkButton
        onClick={() => fields.push({ is_primary: false })}
        title="Add additional project contacts"
      >
        <PlusOutlined /> Add additional project contacts
      </LinkButton>
    </>
  );
};

export const ProjectContacts: FC = () => {
  const dispatch = useDispatch();
  const formValues = useSelector(getFormValues(FORM.ADD_EDIT_PROJECT_SUMMARY));
  const { contacts } = formValues;

  useEffect(() => {
    if (isNil(contacts) || contacts.length === 0) {
      dispatch(arrayPush(FORM.ADD_EDIT_PROJECT_SUMMARY, "contacts", { is_primary: true }));
    }
  }, []);

  return (
    <>
      <Typography.Title level={3}>Project Contacts</Typography.Title>
      <FieldArray name="contacts" contacts={contacts} component={RenderContacts} />
    </>
  );
};

export default ProjectContacts;
