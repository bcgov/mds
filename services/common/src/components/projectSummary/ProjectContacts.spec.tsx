import React from "react";
import { render } from "@testing-library/react";
import { ReduxWrapper } from "@mds/common/tests/utils/ReduxWrapper";
import { PROJECTS, STATIC_CONTENT } from "@mds/common/constants/reducerTypes";
import { FORM } from "../..";
import FormWrapper from "../forms/FormWrapper";
import * as MOCK from "@mds/common/tests/mocks/dataMocks";
import ProjectContacts from "./ProjectContacts";

const initialState = {
  [PROJECTS]: {
    projectSummary: MOCK.PROJECT_SUMMARY,
  },
  [STATIC_CONTENT]: {
    provinceOptions: MOCK.BULK_STATIC_CONTENT_RESPONSE.provinceOptions,
  },
};

describe("ProjectContacts Component", () => {
  it("should render the component with expected fields", () => {
    const { container } = render(
      <ReduxWrapper initialState={initialState}>
        <FormWrapper name={FORM.ADD_EDIT_PROJECT_SUMMARY} initialValues={{}} onSubmit={() => {}}>
          <ProjectContacts />
        </FormWrapper>
      </ReduxWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
