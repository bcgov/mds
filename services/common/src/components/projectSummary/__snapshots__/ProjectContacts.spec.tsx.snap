// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectContacts Component should render the component with expected fields 1`] = `
<div>
  <form
    class="ant-form ant-form-vertical common-form common-form-ADD_EDIT_PROJECT_SUMMARY form-edit"
    id="ADD_EDIT_PROJECT_SUMMARY"
  >
    <h3
      class="ant-typography"
    >
      Project Contacts
    </h3>
    <a
      role="link"
      tabindex="0"
      title="Add additional project contacts"
    >
      <span
        aria-label="plus"
        class="anticon anticon-plus"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="plus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
          />
          <path
            d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
          />
        </svg>
      </span>
       Add additional project contacts
    </a>
  </form>
</div>
`;
